File: 0394-decode-string/0394-decode-string.js
/**
 * @param {string} s
 * @return {string}
 */
var decodeString = function(s) {
    const stack = []
    for (const c of s) {
        if (c === "]") {
            let cur = ""
            while (stack.length && stack.at(-1) !== "[") {
                cur = stack.pop() + cur
            }
            stack.pop()
            let repeat = ""
            while (Number.isInteger(Number(stack.at(-1)))) {
                repeat = stack.pop() + repeat
            }
            stack.push(cur.repeat(repeat.length ? Number(repeat) : 1))
        } else {
            stack.push(c)
        }
    }
    return stack.join("")
};

File: 0394-decode-string/0394-decode-string.py
class Solution:
    def decodeString(self, s: str) -> str:
        stack, i = [], 0
        for c in s:
            if c == "]":
                cur = ""
                while stack and stack[-1] != "[":
                    cur = stack.pop() + cur
                stack.pop()
                k = ""
                while stack and stack[-1].isnumeric():
                    k = stack.pop() + k
                stack.append("".join([cur for _ in range(0, int(k))]))
            else:
                stack.append(c)
        return "".join(stack)